### ISSUE TITLE
Error when using fastspeech2 Tflite model with Android studio

### ISSUE THREAD

**Issue by Zak-SA** [TensorSpeech/TensorFlowTTS#274]
Hi, 

I am getting this error when trying to run my Tflte model on Android studio:

E/InputWorker: Exception: 
    java.lang.IllegalArgumentException: Cannot convert between a TensorFlowLite tensor with type FLOAT32 and a Java object of type [I (which is compatible with the TensorFlowLite type INT32).
        at org.tensorflow.lite.Tensor.throwIfTypeIsIncompatible(Tensor.java:427)
        at org.tensorflow.lite.Tensor.getInputShapeIfDifferent(Tensor.java:287)
        at org.tensorflow.lite.NativeInterpreterWrapper.run(NativeInterpreterWrapper.java:141)
        at org.tensorflow.lite.Interpreter.runForMultipleInputsOutputs(Interpreter.java:347)
        at com.tensorspeech.tensorflowtts.module.FastSpeech2.getMelSpectrogram(FastSpeech2.java:67)
        at com.tensorspeech.tensorflowtts.tts.InputWorker$InputText.proceed(InputWorker.java:86)
        at com.tensorspeech.tensorflowtts.tts.InputWorker$InputText.access$700(InputWorker.java:66)
        at com.tensorspeech.tensorflowtts.tts.InputWorker.lambda$new$0$InputWorker(InputWorker.java:43)
        at com.tensorspeech.tensorflowtts.tts.-$$Lambda$InputWorker$j9Bs3AGMTu4L2S_Dz69XKl4TEps.run(Unknown Source:2)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:457)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)

The code I am using to convert to Tflite using Notebooks is 

!pip install tf-nightly==2.4.0-dev20200630

import numpy as np
import yaml
import tensorflow as tf

from tensorflow_tts.processor import ZAKSpeechProcessor
from tensorflow_tts.processor.ZAKspeech import ZAKSPEECH_SYMBOLS as symbols, _symbol_to_id

from tensorflow_tts.configs import FastSpeechConfig, FastSpeech2Config
from tensorflow_tts.configs import MelGANGeneratorConfig, MultiBandMelGANGeneratorConfig

from tensorflow_tts.models import TFFastSpeech, TFFastSpeech2
from tensorflow_tts.models import TFMelGANGenerator, TFMBMelGANGenerator

from IPython.display import Audio
print(tf.__version__) # check if >= 2.4.0

# initialize MultiBand melgan model
with open('multiband_melgan.v1.yaml') as f:
    mbmelgan_config = yaml.load(f, Loader=yaml.Loader)
mbmelgan_config = MultiBandMelGANGeneratorConfig(**mbmelgan_config["multiband_melgan_generator_params"])
mbmelgan = TFMBMelGANGenerator(config=mbmelgan_config, name='multi_band_melgan_generator')
mbmelgan._build()
mbmelgan.load_weights("multiband_melgan path")

# initialize FastSpeech model.
with open('astspeech2.v1.yaml and path') as f:
    config = yaml.load(f, Loader=yaml.Loader)
config = FastSpeech2Config(**config["fastspeech2_params"])
fastspeech2 = TFFastSpeech2(config=config, name="fastspeech2",
                          enable_tflite_convertible=True)

# Newly added :
fastspeech2._build()
fastspeech2.load_weights("fastspeech2 model")

input_text = "."

processor = processor = ZAKSpeechProcessor(data_dir=None, symbols=symbols, cleaner_names="arabic_cleaners")
input_ids = processor.text_to_sequence(input_text.lower())
input_ids = np.concatenate([input_ids, [len(symbols) - 1]], -1)  # eos.

mel_before, mel_after, duration_outputs, _, _ = fastspeech2.inference(
    input_ids=tf.expand_dims(tf.convert_to_tensor(input_ids, dtype=tf.int32), 0),
    speaker_ids=tf.convert_to_tensor([0], dtype=tf.int32),
    speed_ratios=tf.convert_to_tensor([1.0], dtype=tf.float32),
    f0_ratios=tf.convert_to_tensor([1.0], dtype=tf.float32),
    energy_ratios=tf.convert_to_tensor([1.0], dtype=tf.float32),
)

audio_before = mbmelgan(mel_before)[0, :, 0]
audio_after = mbmelgan(mel_after)[0, :, 0]

#Convert to TFlite

# Concrete Function
fastspeech2_concrete_function = fastspeech2.inference_tflite.get_concrete_function()

converter = tf.lite.TFLiteConverter.from_concrete_functions(
    [fastspeech2_concrete_function]
)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS,
                                       tf.lite.OpsSet.SELECT_TF_OPS]
tflite_model = converter.convert()

# Save the TF Lite model.
with open('fastspeech2_quant.tflite', 'wb') as f:
  f.write(tflite_model)

print('Model size is %f MBs.' % (len(tflite_model) / 1024 / 1024.0) )

# Up to this stage I have no errors and everything work fine.

# Inference from TFLite

import numpy as np
import tensorflow as tf

# Load the TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_path='fastspeech2_quant.tflite')

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Prepare input data.
def prepare_input(input_ids):
  input_ids = tf.expand_dims(tf.convert_to_tensor(input_ids, dtype=tf.int32), 0)
  return (input_ids,
          tf.convert_to_tensor([0], tf.int32),
          tf.convert_to_tensor([1.0], dtype=tf.float32),
          tf.convert_to_tensor([1.0], dtype=tf.float32),
          tf.convert_to_tensor([1.0], dtype=tf.float32))

# Test the model on random input data.
def infer(input_text):
  for x in input_details:
    print(x)
  for x in output_details:
    print(x)
processor = processor = ZAKSpeechProcessor(data_dir=None, symbols=symbols, cleaner_names="arabic_cleaners")
  input_ids = processor.text_to_sequence(input_text.lower())
  interpreter.resize_tensor_input(input_details[0]['index'], 
                                  [1, len(input_ids)])
  interpreter.resize_tensor_input(input_details[1]['index'], 
                                  [1])
  interpreter.resize_tensor_input(input_details[2]['index'], 
                                  [1])
  interpreter.resize_tensor_input(input_details[3]['index'], 
                                  [1])
  interpreter.resize_tensor_input(input_details[4]['index'], 
                                  [1])
  interpreter.allocate_tensors()
  input_data = prepare_input(input_ids)
  for i, detail in enumerate(input_details):
    input_shape = detail['shape']
    interpreter.set_tensor(detail['index'], input_data[i])

  interpreter.invoke()

  # The function `get_tensor()` returns a copy of the tensor data.
  # Use `tensor()` in order to get a pointer to the tensor.
  return (interpreter.get_tensor(output_details[0]['index']),
          interpreter.get_tensor(output_details[1]['index']))

input_text = ""

decoder_output_tflite, mel_output_tflite = infer(input_text)
audio_before_tflite = mbmelgan(decoder_output_tflite)[0, :, 0]
audio_after_tflite = mbmelgan(mel_output_tflite)[0, :, 0]

Audio(data=audio_before_tflite, rate=22050)

Audio(data=audio_after_tflite, rate=22050)

input_text = "."

decoder_output_tflite, mel_output_tflite = infer(input_text)
audio_before_tflite = mbmelgan(decoder_output_tflite)[0, :, 0]
audio_after_tflite = mbmelgan(mel_output_tflite)[0, :, 0]

Audio(data=audio_before_tflite, rate=22050)

Audio(data=audio_after_tflite, rate=22050)

I couldn't inference from Tflite, I already opened another issue for this but still no fix. #237 

I would apreciate the help.

Thanks

**Comment by dathudeptrai** [comment]
I do not really know what is the bug :(. I'm not a person contribute android code so it's hard for me to debug.

**Comment by Zak-SA** [comment]
@dathudeptrai thanks for your reply. The error occurs when loading the Tflite model for fastspeech2 in android.stidio  And the second issue is when inference using Tflite model and I opened another issue earlier but no fix yet. And I noticed that you have discussions related to that issue on Tensorflow github.
I have a question for you about the Tflite model that's included and used with the Android example, what's the code used to convert the fastspeech2 and mbmelgan? 
I noticed the tflite have attention mask when it's removed in the new notebooks. Can we try to use and replicate the same process used to convert the fastspeech2 and mbmelgan used in android example?

**Comment by OscarVanL** [comment]
@Zak-SA Did you ever have success creating `.tflite` models that work with the Android example? I am trying to convert a MB-MelGAN model to work on Android with limited success...

**Comment by mapledxf** [comment]
@Zak-SA please check your Android code.
one of the tflite inputs should be int format instead of float.

**Comment by mapledxf** [comment]
could u show me your android code about inferencing the fastspeech tflie model?

**Comment by Zak-SA** [comment]
@mapledxf thanks.

**Comment by Zak-SA** [comment]
@OscarVanL Yes it's working for me now, to be honest I wasn't the one who solved the issue, I am waiting for the full details on how it was solved and as soon as I get the fix I will share it here. :)

**Comment by mapledxf** [comment]
@dathudeptrai u can assign it to me if there is any question or bug relates to android stuff.

**Comment by dathudeptrai** [comment]
> @dathudeptrai u can assign it to me if there is any question or bug relates to android stuff.

oh, I really appreciate for your support :D.
